src/activity_queue/util.rs

```diff
@@ -0,0 +1,57 @@
+use futures_core::Future;
+use std::{
+    fmt::{Debug, Display},
+    time::Duration,
+};
+
+use tracing::warn;
+
+#[derive(Clone, Copy, Default)]
+pub(super) struct RetryStrategy {
+    /// Amount of time in seconds to back off
+    pub backoff: usize,
+    /// Amount of times to retry
+    pub retries: usize,
+    /// If this particular request has already been retried, you can add an offset here to increment the count to start
+    pub offset: usize,
+    /// Number of seconds to sleep before trying
+    pub initial_sleep: usize,
+}
+
+/// Retries a future action factory function up to `amount` times with an exponential backoff timer between tries
+pub(super) async fn retry<
+    T,
+    E: Display + Debug,
+    F: Future<Output = Result<T, E>>,
+    A: FnMut() -> F,
+>(
+    mut action: A,
+    strategy: RetryStrategy,
+) -> Result<T, E> {
+    let mut count = strategy.offset;
+
+    // Do an initial sleep if it's called for
+    if strategy.initial_sleep > 0 {
+        let sleep_dur = Duration::from_secs(strategy.initial_sleep as u64);
+        tokio::time::sleep(sleep_dur).await;
+    }
```

Comment:
This means that each worker is sleeping separately, right? So a large instance might have thousands of workers active at a time which are in sleep state. I know sleeping is cheap, but wouldnt the large number of active workers still have negative performance effects?

Reply:
I think that keeping the tasks running will also use a lot of memory, because afaik futures are quite big.

Reply:
So it should be more efficient to have a single thread which checks all workers if they are ready to run, and then starts them.

Reply:
I've started a WIP that will:

* Still have the active queue to send things, and if they fail go to the slower retry queue
* Have the slower retry queue check every hour for outstanding requests
* Retries them concurrently (not in parallel), so there is only ever one task for retries
* Sleeps and tries again